### PR TITLE
Modified letsencrypt support to add domain name checking in addition to

### PR DESCRIPTION
**Pull request by neuralp**
hostname checking.  Added necessary tests and renamed original manual
ssl test to a name that supports adding the other SSL tests.

Reference to issue #1274

### CODE REVIEW

**Comment by erik-wramner** [comment]
There are changes not only for letsencrypt but also for the other alternatives. Are there real changes there, or is it only indentation? If there are real changes I need to review more carefully.

There are some places where the if and fi are not aligned, probably you have mixed tabs and spaces. Can you fix that, please? I know that the script is full of indentation problems as is, but at least we can make sure the new parts are aligned.

File: test/mail_ssl_letsencrypt.bats [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    run docker exec mail_lets_hostname /bin/sh -c 'grep -ir "/etc/letsencrypt/live/mail.my-domain.com/" /etc/postfix/main.cf | wc -l'
+    assert_success
+    assert_output 2
+    run docker exec mail_lets_hostname /bin/sh -c 'grep -ir "/etc/letsencrypt/live/mail.my-domain.com/" /etc/dovecot/conf.d/10-ssl.conf | wc -l'
+    assert_success
+    assert_output 2
+}
+
+@test "checking ssl: letsencrypt cert works correctly" {
+    run docker exec mail_lets_domain /bin/sh -c "timeout 1 openssl s_client -connect 0.0.0.0:587 -starttls smtp -CApath /etc/ssl/certs/ | grep 'Verify return code: 10 (certificate has expired)'"
```
Comment (youtous) [inline test/mail_ssl_letsencrypt.bats:57]: Just a thought; could you cover `465` too?

File: test/mail_ssl_letsencrypt.bats [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    docker rm -f mail_lets_hostname
+}
+
+# this test must come first to reliably identify when to run setup_file
+@test "first" {
+    skip 'Starting testing of letsencrypt SSL'
+}
+
+@test "checking ssl: letsencrypt configuration is correct" {
+    run docker exec mail_lets_domain /bin/sh -c 'grep -ir "/etc/letsencrypt/live/my-domain.com/" /etc/postfix/main.cf | wc -l'
```
Comment (youtous) [inline test/mail_ssl_letsencrypt.bats:42]: `postconf` tests are more explicit than counting word

File: test/mail_ssl_letsencrypt.bats [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    assert_success
+    assert_output 2
+    run docker exec mail_lets_domain /bin/sh -c 'grep -ir "/etc/letsencrypt/live/my-domain.com/" /etc/dovecot/conf.d/10-ssl.conf | wc -l'
```
Comment (youtous) [inline test/mail_ssl_letsencrypt.bats:45]: Same for dovecot, `doveconf` is more accurate (https://wiki.dovecot.org/Tools/Doveconf)

File: target/start-mailserver.sh [hunk 4]
```diff
@@ -1042,119 +1042,131 @@ function _setup_ssl() {
...
 
-					notify 'inf' "SSL configured with 'letsencrypt' certificates"
-				else
-					notify 'err' "Key filename not set!"
-				fi
-			else
-				notify 'err' "Cannot access '/etc/letsencrypt/live/"$HOSTNAME"/fullchain.pem'"
-			fi
 		;;
-	"custom" )
```
Comment (youtous) [inline target/start-mailserver.sh:1075]: As @erik-wramner suggested, could you unstage these changes if this is not part of this PR?

File: target/start-mailserver.sh [hunk 5]
```diff
@@ -1042,119 +1042,131 @@ function _setup_ssl() {
...
+            local LETSENCRYPT_DOMAIN=""
+            local LETSENCRYPT_KEY=""
+			
+            # first determine the letsencrypt domain by checking both the full hostname or just the domainname if a SAN is used in the cert
+            if [ -e "/etc/letsencrypt/live/$HOSTNAME/fullchain.pem" ]; then
+                LETSENCRYPT_DOMAIN=$HOSTNAME
+            elif [ -e "/etc/letsencrypt/live/$DOMAINNAME/fullchain.pem" ]; then
+                LETSENCRYPT_DOMAIN=$DOMAINNAME
+			else
+				notify 'err' "Cannot access '/etc/letsencrypt/live/"$HOSTNAME"/fullchain.pem' or '/etc/letsencrypt/live/"$DOMAINNAME"/fullchain.pem'"
```
Comment (youtous) [inline target/start-mailserver.sh:1057]: Shouldn't the script terminate with an error here? `return 1` or something like that?

File: target/start-mailserver.sh [hunk 6]
```diff
@@ -1042,119 +1042,131 @@ function _setup_ssl() {
...
+
+            # then determine the keyfile to use
+			if [ -n "$LETSENCRYPT_DOMAIN" ]; then
+				if [ -e "/etc/letsencrypt/live/$LETSENCRYPT_DOMAIN/privkey.pem" ]; then
+					LETSENCRYPT_KEY="privkey"
+				elif [ -e "/etc/letsencrypt/live/$LETSENCRYPT_DOMAIN/key.pem" ]; then
+					LETSENCRYPT_KEY="key"
 				else
-					notify 'err' "Cannot access '/etc/letsencrypt/live/"$HOSTNAME"/privkey.pem' nor 'key.pem'"
+					notify 'err' "Cannot access '/etc/letsencrypt/live/"$LETSENCRYPT_DOMAIN"/privkey.pem' nor 'key.pem'"
```
Comment (youtous) [inline target/start-mailserver.sh:1067]: Same here

**Comment by neuralp** [comment]
> There are changes not only for letsencrypt but also for the other alternatives. Are there real changes there, or is it only indentation? If there are real changes I need to review more carefully.
> 
> There are so(me places where the if and fi are not aligned, probably you have mixed tabs and spaces. Can you fix that, please? I know that the script is full of indentation problems as is, but at least we can make sure the new parts are aligned.

Since I was working on the SSL function, I figured that I would attempt to fix the indentation since it was all over the place.  I'll revisit this for the function and make sure that all of the tabs get converted to spaces.  Apparently my editor didn't read the .editorconfig.

File: test/mail_ssl_letsencrypt.bats (hunk 2 above)
Comment (neuralp) [reply test/mail_ssl_letsencrypt.bats:42]: To tell you the truth, the tests were already this way (the word count way) so I just duplicated them for the second container. I'll take a look at a more accurate test for this then, not a problem.

File: target/start-mailserver.sh (hunk 4 above)
Comment (neuralp) [reply target/start-mailserver.sh:1075]: Just trying to fix indentation, I will remove and follow the .editorconfig more closely.

File: target/start-mailserver.sh (hunk 5 above)
Comment (neuralp) [reply target/start-mailserver.sh:1057]: Currently if SSL encounters an error, it just leaves the default configuration in place.  It does make sense that if you configured SSL, it would be assumed to be running correctly if no errors were actually returned from the function.  A return 1 sounds look a good thing to implement.

File: test/mail_ssl_letsencrypt.bats (hunk 1 above)
Comment (neuralp) [reply test/mail_ssl_letsencrypt.bats:57]: Could you please clarify what you are referring to? I can't find an issue with that ID or a PR.

Comment (neuralp) [reply test/mail_ssl_letsencrypt.bats:57]: Never mind, I've got it.  Test port 465 as well.

Comment (neuralp) [reply test/mail_ssl_letsencrypt.bats]: 465 is now covered in the tests with the updated pull request

File: test/mail_ssl_letsencrypt.bats (hunk 2 above)
Comment (neuralp) [reply test/mail_ssl_letsencrypt.bats]: tests now use _postconf_ and search for the full expected certificate name

File: test/mail_ssl_letsencrypt.bats (hunk 3 above)
Comment (neuralp) [reply test/mail_ssl_letsencrypt.bats]: same for _doveconf_ which is now used and checked

File: target/start-mailserver.sh (hunk 4 above)
Comment (neuralp) [reply target/start-mailserver.sh:1075]: indentation updates have been corrected and dropped

File: target/start-mailserver.sh (hunk 5 above)
Comment (neuralp) [reply target/start-mailserver.sh:1057]: In addition to the error, return 1 is used after the error.  Upon successfull completion of the letsencrypt portion, return 0 is given.

File: target/start-mailserver.sh (hunk 6 above)
Comment (neuralp) [reply target/start-mailserver.sh:1068]: Resolved with return 1 upon error.

**Comment by neuralp** [comment]
Your suggestions have been taken into account and the PR has been modified to include these suggestions.  Please review at your earliest convenience.

**Comment by erik-wramner** [comment]
Looks good, thanks!
